text/0122-sdk-hub-scope-merge.md

```diff
@@ -0,0 +1,458 @@
+
+```js
+fastify.addHook('preValidation', () => {
+  // This will be applied to all events in the current request
+  getIsolationScope().addEventProcessor(...);
+});
+```
+
+Similar to the global scope, these are also applied to events. However, isolation scopes can be created, either by us internally (the most common scenario), or also by users. The new APIs for this are:
+
+```js
+// Returns the current isolation scope.
+export function getIsolationScope(): Scope;
+
+// Create a new isolation scope for this scope
+// This will NOT make this scope the isolation scope, but will create a new isolation scope (based on the active isolation scope, if one exists)
+scope.isolate();
+
+// Similar to `withScope`, but it forks a new scope AND sets a new isolation scope for this context
+export function withIsolationScope(callback: (scope) => void): void;
+```
+
+You can fetch the current isolation scope via `getIsolationScope()`. You can define a new isolation scope via `scope.isolate()`, which will define a new isolation scope for this scope, and for all scopes that will be forked off this scope. When a client is created & bound, an isolation scope will immediately be created, similar to the global scope for a client.
+
+An isolation scope is attached to the current execution context, similar to the current scope. There is always exactly one current isolation scope. If you call `getIsolationScope()` before a client has been created, a noop isolation scope is returned, which should be merged in once a client is actually created (same as with the global scope).
+
+Similar to the global scope, an isolation scope is always a separate scope, so nothing will inherit off it - except for a potential superseding isolation scope.
+If an isolation scope is created, and there is already an isolation scope in the current execution context, then the new isolation scope should be forked off the previous one (with copy-on-write):
+
+```js
+Sentry.init();
+
+// There is always an isolation scope
+const initialIsolationScope = Sentry.getIsolationScope();
+initialIsolationScope.setTag("tag1", "aa");
+
+// Auto-instrumentation automatically forks an execution context with a new isolation scope
+app.get("/my-route", function () {
+  const isolationScope = Sentry.getIsolationScope();
+  isolationScope !== initialIsolationScope;
+  // Isolation scope inherits from previously current isolation scope
+  isolationScope.getTags(); // -> tag1: aa
+});
+```
+
+Note that in environments where you do not have multiple execution contexts (e.g. Browser JS), there is basically only a single isolation scope that is fundamentally the same as the global scope. In these environments, global & isolation scope can be used completely interchangeably. However, for consistency we still have these concepts everywhere - but users do not need to care about them, and we can reflect this in the SDK specific documentation.
```

Comment:
I don't think adding the isolation scope is a good idea when it's the same as the global scope. People will ask questions and need clarification about what the difference is. If it's the same, I would drop the isolation scope. We can always add it.

Reply:
it's not the same - the global scope will always be the same, while the isolation scope _can_ be changed per context. In environments where there are no execution contexts (e.g. browser or probably mobile), they are more or less the same. But the idea is generally to not really expose this to users too much - users should just use `Sentry.doX` and not `Sentry.getCurrentScope().doX`, and we route it to the correct scope under the hood. that should work just as well in browser or mobile.

So the concept should still be there for consistency sake, and it _does_ work (e.g. you can call it), it's just not commonly useful by default - which is fine, nobody is forced to use it, and users should not have to think about it! _but_ for example looking at browser, if somebody decides to add and setup zone.js or something like this, or browsers eventually add something like execution contexts, we _can_ then also wire this up to work there as well.

Reply:
I'm OK with your proposal.